servicetalk-loadbalancer/src/main/java/io/servicetalk/loadbalancer/RoundRobinLoadBalancer.java

```diff
@@ -150,29 +150,26 @@ public void onNext(final ServiceDiscovererEvent<ResolvedAddress> event) {
                         event);
                 final List<Host<ResolvedAddress, C>> activeAddresses =
                         activeHostsUpdater.updateAndGet(RoundRobinLoadBalancer.this, currentAddresses -> {
-                            final List<Host<ResolvedAddress, C>> refreshedAddresses =
-                                    new ArrayList<Host<ResolvedAddress, C>>(currentAddresses);
                             final ResolvedAddress addr = requireNonNull(event.address());
-                            for (int i = 0; i < refreshedAddresses.size(); i++) {
-                                Host<ResolvedAddress, C> host = refreshedAddresses.get(i);
-                                if (host.address.equals(addr)) {
-                                    if (event.isAvailable()) {
-                                        LOGGER.debug("Address {} added but it already exists.", addr);
-                                        return currentAddresses;
-                                    } else {
-                                        refreshedAddresses.remove(i);
-                                        host.markInactive();
-                                        return refreshedAddresses;
-                                    }
-                                }
-                            }
+                            final List<Host<ResolvedAddress, C>> refreshedAddresses;
                             if (event.isAvailable()) {
+                                refreshedAddresses = new ArrayList<>(currentAddresses.size() + 1);
+                                refreshedAddresses.addAll(currentAddresses);
                                 refreshedAddresses.add(new Host<>(addr));
-                                return refreshedAddresses;
+                            } else if (currentAddresses.isEmpty()) {
+                                refreshedAddresses = emptyList();
                             } else {
-                                LOGGER.debug("Address {} removed but it does not exist.", addr);
-                                return currentAddresses;
+                                refreshedAddresses = new ArrayList<>(currentAddresses.size() - 1);
```

Comment:
`currentAddresses.size() - 1` may give a zero as the result. Consider rewriting in this way:
```java
if (event.isAvailable()) {
    ...
} else if (currentAddresses.size() <= 1) {
    refreshedAddresses = emptyList();
} else {
    refreshedAddresses = new ArrayList<>(currentAddresses.size() - 1);
    ...
}
```

Reply:
hm, may be worth checking that addresses are equal when `currentAddresses.size() == 0`.

Reply:
Not sure if we need to be too paranoid, but here is the version 😄 
```java
final ResolvedAddress addr = requireNonNull(event.address());
if (event.isAvailable()) {
    if (currentAddresses.isEmpty()) {
        return singletonList(new Host<>(addr));
    }
    final List<Host<ResolvedAddress, C>> refreshedAddresses =
            new ArrayList<>(currentAddresses.size() + 1);
    refreshedAddresses.addAll(currentAddresses);
    refreshedAddresses.add(new Host<>(addr));
    return refreshedAddresses;
} else {
    Host<ResolvedAddress, C> host;
    switch (currentAddresses.size()) {
        case 0:
            return emptyList();
        case 1:
            host = (Host<ResolvedAddress, C>) currentAddresses.get(0);
            return addr.equals(host.address) ? emptyList() : singletonList(host);
        case 2:
            host = (Host<ResolvedAddress, C>) currentAddresses.get(0);
            if (addr.equals(host.address)) {
                return singletonList(currentAddresses.get(1));
            }
            host = (Host<ResolvedAddress, C>) currentAddresses.get(1);
            if (addr.equals(host.address)) {
                return singletonList(currentAddresses.get(0));
            }
            return currentAddresses;
        default:
            final List<Host<ResolvedAddress, C>> refreshedAddresses =
                    new ArrayList<>(currentAddresses.size() - 1);
            for (Host<ResolvedAddress, C> hostItem :
                    (List<Host<ResolvedAddress, C>>) currentAddresses) {
                if (hostItem.address.equals(addr)) {
                    hostItem.markInactive();
                } else {
                    refreshedAddresses.add(hostItem);
                }
            }
            return refreshedAddresses;
    }
}
```

Reply:
this code would be a bit more complex in practice because we need to call `markInactive` when removing, and also for the removal of case 2 we need to check if both elements match bcz there maybe duplicates. I will take some of the suggestions (singleton list for adding when empty, return empty list if after removal the array is empty) but keep it simple as the list being empty isn't generally expected (e.g. we wouldn't be able to communicate with servers in that case).